sandbox/cgroup/tracking.go

```diff
@@ -0,0 +1,282 @@
+		return ErrCannotTrackProcess
+	}
+
+	// We ask the kernel for a random UUID. We need one because each transient
+	// scope needs a unique name. The unique name is comprosed of said UUID and
+	// the snap security tag.
+	uuid, err := randomUUID()
+	if err != nil {
+		return err
+	}
+
+	// Enforcing uniqueness is preferred to reusing an existing scope for
+	// simplicity since doing otherwise and joining an existing scope has
+	// limitations:
+	// - the originally started scope must be marked as a delegate, with all
+	//   consequences.
+	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	unitName := fmt.Sprintf("%s.%s.scope", securityTag, uuid)
+
+	pid := osGetpid()
+tryAgain:
+	// Create a transient scope by talking to systemd over DBus.
+	if err := doCreateTransientScope(conn, unitName, pid); err != nil {
+		switch err {
+		case errDBusUnknownMethod:
+			return ErrCannotTrackProcess
+		case errDBusSpawnChildExited:
+			fallthrough
+		case errDBusNameHasNoOwner:
+			// We cannot activate systemd --user for root, try the system bus
+			// as a fallback.
+			if isSessionBus && osGetuid() == 0 {
+				logger.Debugf("cannot activate systemd --user on session bus, falling back to system bus: %s", err)
+				isSessionBus = false
+				conn, err = dbusutil.SystemBus()
+				if err != nil {
+					logger.Debugf("system bus is not available: %s", err)
+					return ErrCannotTrackProcess
+				}
+				logger.Debugf("using system bus now, session bus could not activate systemd --user")
+				goto tryAgain
+			}
+			return ErrCannotTrackProcess
+		}
+		return err
+	}
+	// We may have created a transient scope but due to a kernel design,
+	// in specific situation when we are in a cgroup owned by one user,
+	// and we want to run a process as a different user, *and* systemd is
+	// older than 238, then this can silently fail.
+	//
+	// Verify the effective tracking cgroup and check that our scope name
+	// is contained therein.
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	if !strings.Contains(path, unitName) {
+		logger.Debugf("systemd could not associate process %d with transient scope %s", pid, unitName)
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+// ConfirmSystemdServiceTracking checks if systemd tracks this process as a snap service.
+//
+// Systemd is placing started services, both user and system, into appropriate
+// tracking groups. Given a security tag we can confirm if the current process
+// belongs to such tracking group and thus could be identified by snapd as
+// belonging to a particular snap and application.
+//
+// If the application process is not tracked then ErrCannotTrackProcess is returned.
+func ConfirmSystemdServiceTracking(securityTag string) error {
+	pid := osGetpid()
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	unitName := fmt.Sprintf("%s.service", securityTag)
+	if !strings.Contains(path, unitName) {
```

Comment:
Shouldn't this be `if !strings.HasSuffix(path, unitName)` ?

Reply:
Contains was more conservative in case systemd included something later, after the service we expected to see but I cannot think of a case where this would be required. I've adjusted the code without any test changes required, to be more strict.